src/Symfony/Component/Messenger/Worker.php

```diff
@@ -51,14 +52,23 @@ class Worker
     private $unacks;
 
     /**
-     * @param ReceiverInterface[] $receivers Where the key is the transport name
+     * @param ReceiverInterface[]                                           $receivers       Where the key is the transport name
+     * @param EventDispatcherInterface|SymfonyEventDispatcherInterface|null $eventDispatcher
      */
-    public function __construct(array $receivers, MessageBusInterface $bus, EventDispatcherInterface $eventDispatcher = null, LoggerInterface $logger = null)
+    public function __construct(array $receivers, MessageBusInterface $bus, object $eventDispatcher = null, LoggerInterface $logger = null)
     {
+        if ($eventDispatcher instanceof SymfonyEventDispatcherInterface) {
+            if (class_exists(Event::class)) {
+                $eventDispatcher = LegacyEventDispatcherProxy::decorate($eventDispatcher);
+            }
+        } elseif (null !== $eventDispatcher && !$eventDispatcher instanceof EventDispatcherInterface) {
+            throw new \TypeError(__METHOD__.sprintf(': Parameter $eventDispatcher expected to be either null or an instance of "%s" or "%s", "%s" given.', EventDispatcherInterface::class, SymfonyEventDispatcherInterface::class, get_debug_type($eventDispatcher)));
+        }
+
```

Comment:
This is fugly, but at least it's much clearer what is going on than the big inlined mess before.  I dig it.

For 6.x, since that requires PHP 8, the type hint can just use a union type instead.

Reply:
On 6.x, because we don't support the constracts v1 package anymore, it should be enough to type-hint PSR-14 only.